### ISSUE TITLE
feature request: Add support for Windows Desktop Apps

### ISSUE THREAD

**Issue by kelter-antunes** [logto-io/logto#4995]
### What problem did you meet?
I was setting up a native windows application. For that, I used the Native App template to see if I could get it to work. In theory, everything should work, and the flow will be very similar with the one on mobile native. The problem is that the regex that checks the uri schema protocol is locking it to only work on mobile environments:

![image](https://github.com/logto-io/logto/assets/6965808/4d644beb-5159-4c26-b3a8-2b562a3a3131)

### Describe what you'd like Logto to have
It would be nice to either have a Desktop Native App template, or having the regex changed to allow uri schema protocols like the one in the image (servicestudiox11://auth)

**Comment by darcyYe** [comment]
Thanks for letting us know your use case, will keep you posted once we have any update

**Comment by charIeszhao** [comment]
Yes, currently the regex requires the URI schema contains a `.`. Is that something different for windows desktop apps? Can you still follow this pattern?

**Comment by kelter-antunes** [comment]
I'm not totally sure if there are any specific requirements when registering a custom URI schema protocol on Windows. Looking at the windows default apps, for example, none of them uses the ` . `.
![image](https://github.com/logto-io/logto/assets/6965808/36dc7a87-42eb-4c63-8363-307e8e71357f)

For my specific use case, I want to enable SSO using logto for a web app and also for a Windows app. When SSO is enabled in the desktop app, it will start the SSO process in the browser that then will call back to the desktop app, as you can see:
![image](https://github.com/logto-io/logto/assets/6965808/34c66566-55f8-4a63-b536-4bf408ed4dd2)

Note: I'm not the owner of the desktop app, neither I have control on the protocol registered on Windows. And I assume that most desktop apps will follow the same steps as this one.

**Comment by charIeszhao** [comment]
OK, thanks for the information. I think this is doable. Let me give the feedback to the team and we'll see what we can do.

**Comment by charIeszhao** [comment]
Unfortunately, according to OIDC protocol, non-domain based custom schemes should not be supported. https://datatracker.ietf.org/doc/html/draft-ietf-oauth-native-apps-06#section-7.1.1

<img width="738" alt="image" src="https://github.com/logto-io/logto/assets/12833674/ef32a895-05e2-4056-9ed3-fdb98938933d">

**Comment by charIeszhao** [comment]
Well, after a quick discussion, we decided to remove the strict check from our side. This will most likely be fixed in the next release. Please stay tuned.

**Comment by charIeszhao** [comment]
Sorry for the late update. And I'm afraid we cannot bypass the schema check for these windows native apps, as our underlying library `node-oidc-provider` strictly checks the scheme pattern for native apps and it requires the scheme to have a reverse domain name and at least have a dot.

Source: https://github.com/panva/node-oidc-provider/blob/main/lib/helpers/client_schema.js#L553-L569

So we're sorry but there is nothing we can do for the moment.

**Comment by leeleito** [comment]
@kelter-antunes Did you get anywhere with your desktop working with the native app template? I need to do the same for a WPF app. Any pointers in the right direction would be very handy :)

**Comment by kelter-antunes** [comment]
> @kelter-antunes Did you get anywhere with your desktop working with the native app template? I need to do the same for a WPF app. Any pointers in the right direction would be very handy :)

Unfortunately, I haven't made much progress on this. I was doing some POC experiments with very limited time and need to revisit this topic.

As of now, I think the only way forward is to self-deploy a custom version of Logto, with an adjusted regex to allow URI schema protocols without dots.

**Comment by leeleito** [comment]
This repositry might help you, stumbled across it whilst trying to get the native app to work. Although it's not working yet with Logto I have it working with another provider.

https://github.com/IdentityModel/IdentityModel.OidcClient.Samples/blob/main/Wpf/Wpf/CallbackManager.cs
